### ISSUE TITLE
sm2 verify fail

### ISSUE THREAD

**Issue by chan369** [openssl/openssl#17804]
```
// testverify.cpp : Defines the entry point for the console application.
//

#include "stdafx.h"
#include "testverify.h"

#include <vector>
using namespace std;
#include <conio.h>


#include "SM2X509Helper.h"

#include <WinCrypt.h>
#include <afxmt.h>



#include <openssl/evp.h>
#include <openssl/asn1.h>
#include <openssl/ec.h>
#include <crypto/sm2.h>
#include <openssl/x509.h>
#include <openssl/pem.h>
#include <openssl/bn.h>

#pragma comment(lib,"libcrypto")





const char*g_cert=
	"-----BEGIN CERTIFICATE-----\n"
	"MIIB8TCCAZagAwIBAgIKFkYQASRwUDgWQDAMBggqgRzPVQGDdQUAMGAxCzAJBgNV\n"
	"BAYTAkNOMREwDwYDVQQKDAhEYXpoaWh1aTEMMAoGA1UECwwDRFpIMQswCQYDVQQI\n"
	"DAJTSDERMA8GA1UEBwwIU2hhbmdoYWkxEDAOBgNVBAMMB1RFU1QwMDIwIBcNMjIw\n"
	"MzAxMDIwMjA0WhgPMjEyMjAyMDUwMjAyMDRaMEUxCzAJBgNVBAYTAkNOMTYwNAYD\n"
	"VQQDDC04MDAwMDAwM182ZWVkODZhYi03YTJhLTRjYjItOTIxZS04NzEyMWY4ZWRj\n"
	"NDYwWTATBgcqhkjOPQIBBggqgRzPVQGCLQNCAARpPdQkdVNFkhOy3jVIOSHhrDkj\n"
	"vF6+23e3dqJg91zKtH+K3gTpLM0qPI+3meF5y0nnUtauzYmB3pAvX2HuIKEEo08w\n"
	"TTALBgNVHQ8EBAMCBsAwHQYDVR0OBBYEFMmOvl3ssD9+PDE32Na30U6/8hKmMB8G\n"
	"A1UdIwQYMBaAFKtH63PJZoe86jn3DSyvhE+gjOOmMAwGCCqBHM9VAYN1BQADRwAw\n"
	"RAIgEUEXvobiiDTAjNZz8p0B0dImL1Lh/bu5ZGz84W/KR7sCIEfE5sbaxDNqS2Rg\n"
	"ZLLiIOY49Ns6IZ1c/LENpLzWUiIp\n"
	"-----END CERTIFICATE-----\n";


unsigned char g_hash[]={0xDA,0x50,0x9E,0xFF,0x7E,0xF5,0x32,0xB5,0x8A,0x03,0xC2,0xAE,0xE4,0x4B,0x23,0x94,
	0x2E,0x01,0x18,0x34,0x4F,0x8C,0x8C,0x27,0x62,0x56,0x4A,0x6D,0x67,0x0E,0xD8,0xD2};


unsigned char g_sign[]={0x30,0x45,0x02,0x20,0x58,0x57,0x6E,0x55,0x7E,0xCF,0x14,0x9A,0x95,0xE8,0xB2,0xF6,
	0x50,0x4D,0xCD,0xA0,0x07,0xA5,0x3C,0x9D,0x77,0x67,0x6E,0x85,0x08,0x03,0xEC,0xE5,
	0xDC,0xF9,0x60,0x89,0x02,0x21,0x00,0x8C,0xA2,0x3A,0x72,0x6B,0x5D,0xBA,0xFA,0x94,
	0xDF,0x22,0x89,0x62,0x28,0xD9,0xFE,0xB2,0x5B,0xC0,0x75,0xF8,0x57,0x62,0x3B,0xBC,
	0x61,0x38,0xC0,0x18,0x65,0xD3,0xD3};


X509* loadcert()
{
	X509 * x=NULL;
	BIO*b=BIO_new_mem_buf(g_cert,strlen(g_cert));
	if(b==NULL)
	{
		return NULL;
	}
	x=PEM_read_bio_X509(b,NULL,NULL,NULL);
	if(x==NULL)
	{
		return NULL;
	}
	BIO_free(b);
	return x;
}
bool GetZ(unsigned char *hash,X509* x)
{
	EVP_PKEY* pk=X509_get0_pubkey(x);
	if(pk==NULL)
	{
		return false;
	}
	EC_KEY*key=EVP_PKEY_get0_EC_KEY(pk);
	if(key==NULL)
	{
		return false;
	}
	int nLen = 0;
	unsigned short entlen = 0;
	entlen=(unsigned short)(16*8);
	unsigned char szTmp[1024] = {0};
	szTmp[0] = (entlen >> 8) & 0xFF;
	szTmp[1] = entlen & 0xFF;
	nLen = 2;
	unsigned char bn[1024];

	memcpy(szTmp+nLen, SM2_DEFAULT_USERID, 16);
	nLen += 16;

	const EC_GROUP*grp=EC_KEY_get0_group(key);
	BIGNUM*p=BN_new();
	BIGNUM*a=BN_new();
	BIGNUM*b=BN_new();
	CAutoFreeBIGNUM abn1(p),abn2(a),abn3(b);
	int r=EC_GROUP_get_curve(grp,p,a,b,NULL);
	if(r==0)
	{
		return false;
	}
	r=BN_bn2bin(a,bn);
	memcpy(szTmp+nLen,bn,32);
	nLen+=32;
	r=BN_bn2bin(b,bn);
	memcpy(szTmp+nLen,bn,32);
	nLen+=32;
	const EC_POINT*g=EC_GROUP_get0_generator(grp);
	unsigned char*bufg;
	r=EC_POINT_point2buf(grp,g,POINT_CONVERSION_UNCOMPRESSED,&bufg,NULL);
	memcpy(szTmp+nLen, bufg+1, 64);
	nLen += 64;
	OPENSSL_free(bufg);
	const EC_POINT*g1=EC_KEY_get0_public_key(key);
	unsigned char*bufg1;
	r=EC_POINT_point2buf(grp,g1,POINT_CONVERSION_UNCOMPRESSED,&bufg1,NULL);
	memcpy(szTmp+nLen, bufg+1, 64);
	nLen += 64;
	OPENSSL_free(bufg1);

	unsigned int size;
	r=EVP_Digest(szTmp,nLen,hash,&size,EVP_sm3(),NULL);
	if(r==0)
	{
		return false;
	}
	return true;
}

int sm2_verify(const unsigned char *sign_rom, size_t sign_rom_len,const unsigned char *result, size_t result_len,EC_KEY *ec)
{
	int ret = 1;
	EVP_MD_CTX *evp_md_ctx = NULL;
	EC_KEY *ec_key = ec;
	EVP_PKEY* public_evp_key = NULL;

	if (sign_rom == NULL || sign_rom_len == 0 || result == NULL || result_len == 0) {
		ret = -1;
		goto finish;
	}
	public_evp_key = EVP_PKEY_new();
	if (public_evp_key == NULL) {
		ret = -1;
		goto finish;
	}
	ret = EVP_PKEY_set1_EC_KEY(public_evp_key, ec_key);
	if (ret != 1) {
		ret = -1;
		goto finish;
	}
	ret = EVP_PKEY_set_alias_type(public_evp_key, EVP_PKEY_SM2);
	if (ret != 1) {
		ret = -1;
		goto finish;
	}
	evp_md_ctx = EVP_MD_CTX_new();
	if (evp_md_ctx == NULL) {
		ret = -1;
		goto finish;
	}
	EVP_MD_CTX_init(evp_md_ctx);
	ret = EVP_VerifyInit_ex(evp_md_ctx, EVP_sm3(), NULL);
	if (ret != 1) {
		ret = -1;
		goto finish;
	}
	ret = EVP_VerifyUpdate(evp_md_ctx, result, result_len);
	if (ret != 1) {
		printf("EVP_VerifyUpdate failed, ret = %d\n", ret);
		ret = -1;
		goto finish;
	}

	ret = EVP_VerifyFinal(evp_md_ctx, sign_rom, (unsigned int)sign_rom_len, public_evp_key);
	if (ret != 1) {
		ret = -1;
		goto finish;
	}
	ret = 1;
finish:
	if (evp_md_ctx != NULL)
		EVP_MD_CTX_free(evp_md_ctx);
	if (public_evp_key != NULL)
		EVP_PKEY_free(public_evp_key);

	return ret;
}

bool verify()
{
	X509 * x=NULL;
	x=loadcert();
	if(x==NULL)
	{
		return false;
	}
	unsigned char hash[32];
	if(GetZ(hash,x)==false)
	{
		return false;
	}
	
	unsigned char sm3[1024];
	unsigned int size=1024;

	unsigned char d2[4096];
	memcpy(d2,hash,32);
	memcpy(d2+32,g_hash,sizeof(g_hash));

	EVP_PKEY* pk=X509_get0_pubkey(x);
	if(pk==NULL)
	{
		return false;
	}

	EC_KEY*key=EVP_PKEY_get0_EC_KEY(pk);
	if(key==NULL)
	{
		return false;
	}
    int r=sm2_verify(g_sign,sizeof(g_sign),d2,sizeof(g_hash)+32,key);
    bool b=(r==1);
	X509_free(x);
	return b;
}


DWORD WINAPI testproc(void*)
{
	if(verify()==false)
	{
		printf("multithread test fail\n");
	}
	return 0;
}





#ifdef _DEBUG
#define new DEBUG_NEW
#endif


// The one and only application object

CWinApp theApp;

using namespace std;

int _tmain(int argc, TCHAR* argv[], TCHAR* envp[])
{
	int nRetCode = 0;

	HMODULE hModule = ::GetModuleHandle(NULL);

	if (hModule != NULL)
	{
		// initialize MFC and print and error on failure
		if (!AfxWinInit(hModule, NULL, ::GetCommandLine(), 0))
		{
			// TODO: change error code to suit your needs
			_tprintf(_T("Fatal Error: MFC initialization failed\n"));
			nRetCode = 1;

		}
		else
		{
			int count=10;

			//test
			for(int i=0;i<count;i++)
			{
				if(verify()==false)
				{
					printf("single thread test fail\n");
				}
			}



			vector<HANDLE> vt;
			for(int i=0;i<count;i++)
			{
				vt.push_back(CreateThread(NULL,0,testproc,0,0,0));
			}
			for(int i=0;i<count;i++)
			{
				WaitForSingleObject(vt[i],INFINITE);
				CloseHandle(vt[i]);
			}
			printf("finish\n");



			_getch();
			return 0;
		}
	}
	else
	{
		// TODO: change error code to suit your needs
		_tprintf(_T("Fatal Error: GetModuleHandle failed\n"));
		nRetCode = 1;
	}

	return nRetCode;
}

   ```
   ```
test1 always success.
but  int test2's loop,sometimes suceed,,sometimes fail.
if i add a criticalsection to protect verify(),it's hard to fail,but if i change loop variant count to 400, There is a small probability of failure
what's the problem?
   ```

**Comment by chan369** [comment]
my openssl version is 1.1.1m

**Comment by zzl360** [comment]
do you mean,when use multithread mode to verify sm2 cert, sometimes it may fail?
I think you should add more log in your code,and print every return value to check which is failed?

**Comment by chan369** [comment]
> ```
> #include <crypto/sm2.h>
> ```
> 
> Don't do this! The functions in that header file are _internal_ to OpenSSL. You must not use them in your own applications. You need to use the EVP_* functions instead for doing SM2.

i don't use internal function to verify,i use evp_* functions;

**Comment by chan369** [comment]
> do you mean,when use multithread mode to verify sm2 cert, sometimes it may fail? I think you should add more log in your code,and print every return value to check which is failed?

test2 always fail  when  i call EVP_VerifyFinal

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of the 3.4 dev cycle, barring further input
